src/nvim/buffer.c

```diff
@@ -3272,212 +3433,362 @@ build_stl_str_hl (
-        s = item[0].start;
-        l = 0;
-      }
     }
 
-    if (width - vim_strsize(s) >= maxwidth) {
-      /* Truncation mark is beyond max length */
+    // If the truncation point we found is beyond the maximum
+    // length of the string, truncate the end of the string.
+    if (width - vim_strsize(trunc_p) >= maxwidth) {
+      // If we are using a multi-byte encoding, walk from the beginning of the
+      // string to find the last character that will fit.
       if (has_mbyte) {
-        s = out;
+        trunc_p = out;
         width = 0;
         for (;; ) {
-          width += ptr2cells(s);
+          width += ptr2cells(trunc_p);
           if (width >= maxwidth)
             break;
-          s += (*mb_ptr2len)(s);
+
+          // Note: Only advance the pointer if the next
+          //       character will fit in the available output space
+          trunc_p += (*mb_ptr2len)(trunc_p);
         }
-        /* Fill up for half a double-wide character. */
-        while (++width < maxwidth)
-          *s++ = fillchar;
-      } else
-        s = out + maxwidth - 1;
-      for (l = 0; l < itemcnt; l++)
-        if (item[l].start > s)
+        // Fill up for half a double-wide character.
+        // XXX : This seems impossible given the exit condition above?
+        while (++width < maxwidth) {
+          *trunc_p++ = fillchar;
+        }
+
+      // Otherwise put the truncation point at the end, leaving enough room
+      // for a single-character truncation marker
+      } else {
+        trunc_p = out + maxwidth - 1;
+      }
+
+      // Ignore any items in the statusline that occur after
+      // the truncation point
+      for (int i = 0; i < itemcnt; i++) {
+        if (item[i].start > trunc_p) {
+          itemcnt = i;
           break;
-      itemcnt = l;
-      *s++ = '>';
-      *s = 0;
+        }
+      }
+
+      // Truncate the output
+      *trunc_p++ = '>';
+      *trunc_p = 0;
+
+    // Truncate at the truncation point we found
     } else {
+      // { Determine how many bytes to remove
+      long trunc_len;
       if (has_mbyte) {
-        n = 0;
+        trunc_len = 0;
         while (width >= maxwidth) {
-          width -= ptr2cells(s + n);
-          n += (*mb_ptr2len)(s + n);
+          width     -= ptr2cells(trunc_p + trunc_len);
+          trunc_len += (*mb_ptr2len)(trunc_p + trunc_len);
         }
-      } else
-        n = width - maxwidth + 1;
-      p = s + n;
-      STRMOVE(s + 1, p);
-      *s = '<';
+      } else {
+        // Truncate an extra character so we can insert our `<`.
+        trunc_len = (width - maxwidth) + 1;
+      }
+      // }
+
+      // { Truncate the string
+      char_u *trunc_end_p = trunc_p + trunc_len;
+      STRMOVE(trunc_p + 1, trunc_end_p);
 
-      /* Fill up for half a double-wide character. */
+      // Put a `<` to mark where we truncated at
+      *trunc_p = '<';
+
+      // Advance the pointer to the end of the string
+      trunc_p = trunc_p + STRLEN(trunc_p);
```

Comment:
This line also needs to check `width+1<maxwidth` before changing `trunc_p`. Otherwise too long tablines could crash nvim.